# 지표 지도 그리기

## Funnel & North Star Framework

### 1. 분석 목적

고객이 프로모션을 수신한 뒤 열람하고 조건을 달성하기까지의 과정을 퍼널로 구성한다.

각 단계의 전환율을 확인하고, 프로모션 설계와 고객 특성에 따라 성과가 어떻게 달라지는지 분석한다.

---

## 2. North Star Metric

### 인센티브형 프로모션의 열람 후 완료율

> 열람 후 유효기간 내 완료된 BOGO·할인형 프로모션 수  
> ÷ 열람된 BOGO·할인형 프로모션 수 × 100

열람 후 완료율은 고객이 프로모션을 확인한 후 실제로 조건을 달성한 정도를 보여준다.

정보 제공형 프로모션은 보상과 달성 조건이 없어 완료 이벤트가 발생하지 않으므로 North Star Metric 계산에서 제외한다.

---

## 3. 프로모션 퍼널

| 퍼널 단계 | 지표 | 계산 기준 | 의미 |
|---|---|---|---|
| 수신 | 프로모션 수신 건수 | `offer received` 기록 수 | 고객에게 발송된 프로모션 규모 |
| 수신 → 열람 | 수신 후 열람률 | 열람된 프로모션 수 ÷ 수신된 프로모션 수 × 100 | 수신한 프로모션을 확인한 비율 |
| 열람 → 완료 | 열람 후 완료율 | 열람 후 완료된 인센티브형 프로모션 수 ÷ 열람된 인센티브형 프로모션 수 × 100 | 프로모션을 확인한 후 조건을 달성한 비율 |
| 수신 → 완료 | 프로모션 완료율 | 완료된 인센티브형 프로모션 수 ÷ 수신된 인센티브형 프로모션 수 × 100 | 수신한 프로모션 중 최종 완료된 비율 |
| 결제 | 결제금액 | `transaction`의 `amount` 합계 | 고객의 전체 구매 규모를 보여주는 보조 지표 |

인센티브형 프로모션은 BOGO와 할인형 프로모션을 의미한다.

---

## 4. 가설-지표 매핑

| 구분 | 가설 | 퍼널 단계 | 측정 지표 | 비교 기준 | 사용 데이터 | 측정 가능 여부 |
|---|---|---|---|---|---|---|
| 프로모션 유형 | 프로모션 유형에 따라 열람률에 차이가 있을 것이다 | 수신 → 열람 | 수신 후 열람률 | BOGO·할인·정보 제공형 | portfolio, transcript | 가능 |
| 프로모션 유형 | 할인형은 BOGO보다 완료율이 높을 것이다 | 수신 → 완료 | 프로모션 완료율 | 할인형·BOGO | portfolio, transcript | 가능 |
| 보상 | 보상 수준이 높을수록 완료율이 높을 것이다 | 열람 → 완료 | 열람 후 완료율 | 보상 수준 | portfolio, transcript | 가능 |
| 유효기간 | 유효기간이 길수록 완료율이 높을 것이다 | 열람 → 완료 | 열람 후 완료율 | 유효기간 | portfolio, transcript | 가능 |
| 난이도 | 달성 난이도가 높을수록 완료율이 낮을 것이다 | 열람 → 완료 | 열람 후 완료율 | 최소 결제 조건 | portfolio, transcript | 가능 |
| 채널 | 발송 채널 조합에 따라 열람률과 완료율에 차이가 있을 것이다 | 수신 → 열람, 열람 → 완료 | 수신 후 열람률, 열람 후 완료율 | 채널 포함 여부 및 조합 | portfolio, transcript | 제한적으로 가능 |
| 성별 | 여성은 남성보다 완료율이 높을 것이다 | 열람 → 완료 | 열람 후 완료율 | 여성·남성 | profile, transcript | 가능 |
| 소득 | 소득이 높은 고객은 완료율이 높을 것이다 | 열람 → 완료 | 열람 후 완료율 | 소득 구간 | profile, transcript | 가능 |
| 연령 | 연령이 높은 고객은 완료율이 낮을 것이다 | 열람 → 완료 | 열람 후 완료율 | 연령대 | profile, transcript | 가능 |

---

## 5. 지표 계산 기준

### 수신 후 열람률

> 열람된 프로모션 수 ÷ 수신된 프로모션 수 × 100

BOGO·할인·정보 제공형을 모두 포함해 계산할 수 있다.

### 프로모션 완료율

> 완료된 인센티브형 프로모션 수 ÷ 수신된 인센티브형 프로모션 수 × 100

완료 조건이 있는 BOGO와 할인형만 포함한다.

### 열람 후 완료율

> 열람 후 완료된 인센티브형 프로모션 수 ÷ 열람된 인센티브형 프로모션 수 × 100

열람 시점보다 완료 시점이 늦고, 프로모션 유효기간 안에 완료된 경우만 포함한다.

---

## 6. 비교 구간 초안

가설 검정에 사용할 구간은 팀 논의를 통해 최종 확정한다.

현재 데이터에서 확인한 기준값은 다음과 같다.

| 항목 | 확인된 기준값 |
|---|---:|
| 보상 평균 | 4.2 |
| 유효기간 평균 | 6.5일 |
| 난이도 평균 | 7.7 |
| 고객 평균 나이 | 약 54.39세 |
| 고객 소득 중앙값 | 64,000달러 |
| 고객 소득 평균 | 약 65,405달러 |

- 연령을 평균 기준으로 구분하면 54세 이하와 55세oria 이상으로 나눌 수 있다.
- 소득 8만 달러 이상을 고소득 고객으로 설정할 경우, 데이터BELMelissa 평균이 아니라 팀이 정한 구간임을 명시한다.

---

## 7. 측정 시 주의사항

- 동일 고객이 같은 프로모션을 여러 번 받을 수 있으므로 고객·프로모션·수신 시점을 기준으로 각 프로모션 수신 건을 구분해야 한다.
- 열람과 완료는 해당 프로모션의 수신 이후이면서 유효기간 안에 발생한 경우만 연결해야 한다.
- 정보 제공형 프로모션은 완료 조건이 없으므로 열람률까지만 평가한다.
- 여러 채널이 하나의 프로모션에 함께 포함되어 있어 개별 채널의 독립적인 효과를 확정하기 어렵다.
- 채널 분석은 채널의 인과효과가 아니라 채널 포함 여부 또는 채널 조합과 성과의 연관성으로 해석한다.
- 고객정보가 누락되고 나이가 118세로 기록된 2,175명은 고객 특성별 분석에서 별도로 처리한다.
- 결제 기록에는 프로모션 ID가 없으므로 특정 결제를 특정 프로모션의 성과로 직접 귀속할 수 없다.
- 프로모션 설계는 10개뿐이므로 보상·기간·난이도·채널 각각의 독립적인 효과를 강하게 단정하지 않는다.

---

## 8. 현재 측정하기 어려운 지표

| 지표 | 측정이 어려운 이유 | 추가로 필요한 데이터 |
|---|---|---|
| 개별 채널의 직접 효과 | 실제로 어떤 채널에서 열람했는지 알 수 없음 | 채널별 노출·클릭 로그 |
| 프로모션의 정확한 구매 유발 효과 | 결제 기록에 프로모션 ID가 없음 | 결제별 프로모션 연결 정보 또는 대조군 |
| 정확한 ROI | 프로모션 운영비와 상품 원가 정보가 없음 | 캠페인 비용·상품 원가·마진 데이터 |
| 장기 재구매 효과 | 관찰기간이 약 30일로 짧음 | 장기간 고객 행동 기록 |

# 지표 계산

## 1. 데이터 불러오기

In [1]:
import pandas as pd
from pathlib import Path

DATA_DIR = Path("raw")

portfolio = pd.read_json(
    DATA_DIR / "portfolio.json",
    lines=True
)

profile = pd.read_json(
    DATA_DIR / "profile.json",
    lines=True
)

transcript = pd.read_json(
    DATA_DIR / "transcript.json",
    lines=True
)

print(portfolio.shape)
print(profile.shape)
print(transcript.shape)

(10, 6)
(17000, 5)
(306534, 4)


## 2. 프로모션 ID 추출

In [2]:
transcript_work = transcript.copy()

transcript_work["offer_id"] = transcript_work["value"].apply(
    lambda x: x.get("offer id", x.get("offer_id"))
)

display(
    transcript_work[
        ["person", "event", "offer_id", "time"]
    ].head()
)

,person,event,offer_id,time
0,78afa995795e4d85b5d9ceeca43f5fef,offer received,9b98b8c7a33c4b65b9aebfe6a799e6d9,0
1,a03223e636434f42ac4c3df47e8bac43,offer received,0b1e1539f2cc45b7b9fa7c272da2e1d7,0
2,e2127556f4f64592b11af22de27a7932,offer received,2906b810c7d4411798c6938adc9daaa5,0
3,8ec6ce2a7e7949b1bf142def7d0e0586,offer received,fafdcd668e3743c1bb461111dcafc2a4,0
4,68617ca6246f4fbc85e91a2a49552598,offer received,4d5c57ea9a6940dd891ad53e9dbe8da0,0


In [3]:
offer_id_check = (
    transcript_work
    .groupby("event")["offer_id"]
    .agg(
        전체_기록="size",
        ID_존재="count"
    )
)

display(offer_id_check)

,전체_기록,ID_존재
event,,
offer completed,33579,33579
offer received,76277,76277
offer viewed,57725,57725
transaction,138953,0


### 추출 결과

- 프로모션 수신 76,277건, 열람 57,725건, 완료 33,579건에서 프로모션 ID가 모두 정상적으로 추출되었다.
- 결제 138,953건에는 원본 데이터상 프로모션 ID가 없으므로 `offer_id`가 결측치로 남는 것이 정상이다.
- 프로모션 관련 기록에서 ID가 누락된 경우는 확인되지 않았다.

## 3. 수신 건별 열람·완료 연결

In [4]:
# 프로모션 유효기간을 시간 단위로 변환
duration_hours = (
    portfolio.set_index("id")["duration"] * 24
).to_dict()

# 결제를 제외한 프로모션 행동만 선택
offer_events = transcript_work.loc[
    transcript_work["event"] != "transaction",
    ["person", "event", "time", "offer_id"]
].copy()

# 같은 시간에는 수신 → 열람 → 완료 순서로 처리
event_order = {
    "offer received": 0,
    "offer viewed": 1,
    "offer completed": 2
}

offer_events["event_order"] = (
    offer_events["event"].map(event_order)
)

offer_events = offer_events.sort_values(
    ["person", "offer_id", "time", "event_order"]
)

offer_instances = []

for (person, offer_id), group in offer_events.groupby(
    ["person", "offer_id"],
    sort=False
):
    received_instances = []

    for row in group.itertuples(index=False):

        if row.event == "offer received":
            received_instances.append({
                "customer_id": person,
                "offer_id": offer_id,
                "received_time": row.time,
                "expiry_time": (
                    row.time + duration_hours[offer_id]
                ),
                "viewed_time": None,
                "completed_time": None
            })

        elif row.event == "offer viewed":
            candidates = [
                instance
                for instance in received_instances
                if (
                    instance["received_time"]
                    <= row.time
                    <= instance["expiry_time"]
                    and instance["viewed_time"] is None
                )
            ]

            if candidates:
                latest_instance = max(
                    candidates,
                    key=lambda x: x["received_time"]
                )
                latest_instance["viewed_time"] = row.time

        elif row.event == "offer completed":
            candidates = [
                instance
                for instance in received_instances
                if (
                    instance["received_time"]
                    <= row.time
                    <= instance["expiry_time"]
                    and instance["completed_time"] is None
                )
            ]

            if candidates:
                latest_instance = max(
                    candidates,
                    key=lambda x: x["received_time"]
                )
                latest_instance["completed_time"] = row.time

    offer_instances.extend(received_instances)

offer_funnel = pd.DataFrame(offer_instances)

print("생성된 프로모션 수신 건:", len(offer_funnel))

생성된 프로모션 수신 건: 76277


### 프로모션 수신 단위 생성 결과

고객별 프로모션 성과를 계산하기 위해 행동 기록을 프로모션 수신 건 단위로 재구성했다.

- 원본 프로모션 수신 기록: 76,277건
- 생성된 프로모션 수신 단위: 76,277건
- 누락되거나 추가된 수신 기록: 0건

각 수신 건에는 고객 ID, 프로모션 ID, 수신 시점과 유효기간을 부여했다. 이후 유효기간 안에 발생한 열람 및 완료 기록을 해당 수신 건에 연결한다.

동일 고객이 동일한 프로모션을 여러 번 받은 경우에는 행동 발생 시점을 기준으로 유효한 가장 최근 수신 건에 연결하는 잠정 기준을 사용했다. 최종 분석 전 연결되지 않은 열람·완료 기록과 중복 기록을 추가로 점검해야 한다.

## 4. 프로모션 정보 연결

In [5]:
offer_funnel = offer_funnel.merge(
    portfolio[
        [
            "id",
            "offer_type",
            "reward",
            "difficulty",
            "duration",
            "channels"
        ]
    ],
    left_on="offer_id",
    right_on="id",
    how="left"
).drop(columns="id")

offer_funnel["viewed"] = (
    offer_funnel["viewed_time"].notna()
)

offer_funnel["completed"] = (
    offer_funnel["completed_time"].notna()
)

offer_funnel["viewed_before_completed"] = (
    offer_funnel["viewed"]
    & offer_funnel["completed"]
    & (
        offer_funnel["viewed_time"]
        <= offer_funnel["completed_time"]
    )
)

offer_funnel["completed_without_prior_view"] = (
    offer_funnel["completed"]
    & ~offer_funnel["viewed_before_completed"]
)

display(offer_funnel.head())

,customer_id,offer_id,received_time,expiry_time,viewed_time,completed_time,offer_type,reward,difficulty,duration,channels,viewed,completed,viewed_before_completed,completed_without_prior_view
0,0009655768c64bdeb2e877511632db8f,2906b810c7d4411798c6938adc9daaa5,576,744,NaN,576.0,discount,2,10,7,"[web, email, mobile]",False,True,False,True
1,0009655768c64bdeb2e877511632db8f,3f207df678b143eea3cee63160fa8bed,336,432,372.0,NaN,informational,0,0,4,"[web, email, mobile]",True,False,False,False
2,0009655768c64bdeb2e877511632db8f,5a8bc65990b245e5a138643cd4eb9837,168,240,192.0,NaN,informational,0,0,3,"[email, mobile, social]",True,False,False,False
3,0009655768c64bdeb2e877511632db8f,f19421c1d4aa40978ebb69ca19b0e20d,408,528,456.0,414.0,bogo,5,5,5,"[web, email, mobile, social]",True,True,False,True
4,0009655768c64bdeb2e877511632db8f,fafdcd668e3743c1bb461111dcafc2a4,504,744,540.0,528.0,discount,2,10,10,"[web, email, mobile, social]",True,True,False,True


In [6]:
print("[프로모션 정보 결측치]")
print(
    offer_funnel[
        [
            "offer_type",
            "reward",
            "difficulty",
            "duration",
            "channels"
        ]
    ].isna().sum()
)

print("\n[만료시간 계산 확인]")
expiry_check = (
    offer_funnel["expiry_time"]
    == offer_funnel["received_time"]
    + offer_funnel["duration"] * 24
)

print("만료시간이 잘못 계산된 건수:", (~expiry_check).sum())

print("\n[정보 제공형 완료 확인]")
informational_completed = (
    (offer_funnel["offer_type"] == "informational")
    & offer_funnel["completed"]
).sum()

print("정보 제공형 완료 건수:", informational_completed)

[프로모션 정보 결측치]
offer_type    0
reward        0
difficulty    0
duration      0
channels      0
dtype: int64

[만료시간 계산 확인]
만료시간이 잘못 계산된 건수: 0

[정보 제공형 완료 확인]
정보 제공형 완료 건수: 0


### 프로모션 정보 및 행동 순서 연결

 ### 확인 목적

 프로모션 수신 건에 유형, 보상, 달성 조건, 유효기간과 발송 채널을 연결하고, 열람과 완료가 발생한 순서를 구분했다.

 ### 생성된 주요 변수

 - `viewed`: 유효기간 안에 프로모션을 열람했는지
 - `completed`: 유효기간 안에 프로모션 조건을 달성했는지
 - `viewed_before_completed`: 프로모션을 열람한 뒤 완료했는지
 - `completed_without_prior_view`: 프로모션을 열람하기 전에 완료했는지

 ### 해석

 일부 고객은 프로모션을 열람하기 전에 완료 조건을 달성했다. 따라서 단순히 열람 기록과 완료 기록이 모두 존재하는지만 확인하면 안 되고, 
  열람과 완료의 시간 순서를 함께 고려해야 한다.

 정보 제공형 프로모션은 완료 조건이 없으므로 완료 여부가 아닌 열람 여부를 중심으로 평가한다.

## 5. 전체 퍼널 지표 계산

In [7]:
incentive_mask = offer_funnel["offer_type"].isin(
    ["bogo", "discount"]
)

all_received = len(offer_funnel)
all_viewed = offer_funnel["viewed"].sum()

incentive_received = incentive_mask.sum()
incentive_viewed = (
    incentive_mask & offer_funnel["viewed"]
).sum()

incentive_completed = (
    incentive_mask & offer_funnel["completed"]
).sum()

viewed_then_completed = (
    incentive_mask
    & offer_funnel["viewed_before_completed"]
).sum()

completed_without_view = (
    incentive_mask
    & offer_funnel["completed_without_prior_view"]
).sum()

print("[전체 프로모션]")
print("수신 건수:", all_received)
print("열람 건수:", all_viewed)
print("수신 후 열람률:", round(all_viewed / all_received * 100, 2), "%")

print("\n[BOGO·할인형 프로모션]")
print("수신 건수:", incentive_received)
print("열람 건수:", incentive_viewed)
print("완료 건수:", incentive_completed)

print(
    "수신 후 열람률:",
    round(incentive_viewed / incentive_received * 100, 2),
    "%"
)

print(
    "수신 후 완료율:",
    round(incentive_completed / incentive_received * 100, 2),
    "%"
)

print(
    "열람 후 완료율:",
    round(viewed_then_completed / incentive_viewed * 100, 2),
    "%"
)

print(
    "완료 중 사전 열람이 없었던 비율:",
    round(completed_without_view / incentive_completed * 100, 2),
    "%"
)

[전체 프로모션]
수신 건수: 76277
열람 건수: 56567
수신 후 열람률: 74.16 %

[BOGO·할인형 프로모션]
수신 건수: 61042
열람 건수: 46620
완료 건수: 33549
수신 후 열람률: 76.37 %
수신 후 완료율: 54.96 %
열람 후 완료율: 50.53 %
완료 중 사전 열람이 없었던 비율: 29.79 %


## 전체 프로모션 퍼널 지표

### 계산 결과

| 지표 | 결과 |
|---|---:|
| 전체 프로모션 수신 건수 | 76,277건 |
| 전체 프로모션 열람 건수 | 56,567건 |
| 전체 수신 후 열람률 | 74.16% |
| BOGO·할인형 수신 건수 | 61,042건 |
| BOGO·할인형 열람 건수 | 46,620건 |
| BOGO·할인형 완료 건수 | 33,549건 |
| BOGO·할인형 수신 후 열람률 | 76.37% |
| BOGO·할인형 수신 후 완료율 | 54.96% |
| BOGO·할인형 열람 후 완료율 | 50.53% |
| 완료 중 사전 열람이 없었던 비율 | 29.79% |

### 해석

전체 프로모션 수신 건의 74.16%가 유효기간 안에 열람됐다. BOGO·할인형 프로모션의 수신 후 열람률은 76.37%로 나타났다.

BOGO·할인형을 열람한 수신 건 중 열람 이후 완료된 비율은 50.53%였다. 완료 건의 29.79%는 프로모션을 열람하기 전에 완료된 기록이었다.

수신 후 완료율은 참고값으로만 사용하고, 핵심 지표는 수신 후 열람률과 열람 후 완료율로 설정한다.

사전 열람 없이 완료한 고객을 프로모션 없이도 구매할 고객이라고 확정할 수는 없다. 이는 열람과 완료의 시간 순서에 관한 관측 결과로 해석한다.

또한 관찰 종료 시점 이후까지 유효기간이 남은 프로모션이 있으므로, 최종 지표 확정 전 관찰기간이 완전히 확보된 수신 건만 이용한 결과와 비교해야 한다.

### 6. 관찰기간·미연결 이벤트·중복 기록 점검

In [8]:
# 6. 관찰기간·미연결 이벤트·중복 기록 점검

observation_end = transcript_work["time"].max()

# 유효기간 전체가 관찰기간 안에 들어온 수신 건
fully_observed_mask = (
    offer_funnel["expiry_time"] <= observation_end
)

fully_observed = offer_funnel.loc[
    fully_observed_mask
].copy()

incentive_mask = fully_observed["offer_type"].isin(
    ["bogo", "discount"]
)

incentive_observed = fully_observed.loc[
    incentive_mask
].copy()


# 1) 관찰기간 점검
print("[관찰기간 점검]")
print("관찰 종료 시간:", observation_end)
print("전체 수신 건:", len(offer_funnel))
print("유효기간 전체가 관찰된 수신 건:", len(fully_observed))
print(
    "관찰기간이 부족한 수신 건:",
    (~fully_observed_mask).sum()
)


# 2) 관찰기간이 확보된 데이터의 퍼널 지표
all_view_rate = (
    fully_observed["viewed"].mean() * 100
)

incentive_view_rate = (
    incentive_observed["viewed"].mean() * 100
)

incentive_completion_rate = (
    incentive_observed["completed"].mean() * 100
)

viewed_incentive = incentive_observed.loc[
    incentive_observed["viewed"]
]

view_to_completion_rate = (
    viewed_incentive["viewed_before_completed"].mean()
    * 100
)

completed_incentive = incentive_observed.loc[
    incentive_observed["completed"]
]

no_prior_view_share = (
    completed_incentive[
        "completed_without_prior_view"
    ].mean()
    * 100
)

print("\n[관찰기간이 확보된 수신 건 기준]")
print("전체 수신 후 열람률:", round(all_view_rate, 2), "%")
print(
    "BOGO·할인형 수신 후 열람률:",
    round(incentive_view_rate, 2),
    "%"
)
print(
    "BOGO·할인형 수신 후 완료율:",
    round(incentive_completion_rate, 2),
    "%"
)
print(
    "BOGO·할인형 열람 후 완료율:",
    round(view_to_completion_rate, 2),
    "%"
)
print(
    "완료 중 사전 열람이 없었던 비율:",
    round(no_prior_view_share, 2),
    "%"
)


# 3) 원본 이벤트와 연결된 이벤트 비교
raw_viewed = (
    transcript_work["event"] == "offer viewed"
).sum()

raw_completed = (
    transcript_work["event"] == "offer completed"
).sum()

linked_viewed = offer_funnel["viewed"].sum()
linked_completed = offer_funnel["completed"].sum()

print("\n[미연결 이벤트 점검]")
print("원본 열람 이벤트:", raw_viewed)
print("수신 건에 연결된 열람:", linked_viewed)
print("연결되지 않은 열람:", raw_viewed - linked_viewed)

print("원본 완료 이벤트:", raw_completed)
print("수신 건에 연결된 완료:", linked_completed)
print("연결되지 않은 완료:", raw_completed - linked_completed)


# 4) 중복 의심 기록 확인
duplicate_mask = (
    transcript.astype(str).duplicated()
)

print("\n[중복 의심 기록]")
print("전체 중복 의심 기록:", duplicate_mask.sum())

display(
    transcript.loc[
        duplicate_mask,
        "event"
    ].value_counts().to_frame("중복 의심 건수")
)

[관찰기간 점검]
관찰 종료 시간: 714
전체 수신 건: 76277
유효기간 전체가 관찰된 수신 건: 66041
관찰기간이 부족한 수신 건: 10236

[관찰기간이 확보된 수신 건 기준]
전체 수신 후 열람률: 74.83 %
BOGO·할인형 수신 후 열람률: 77.69 %
BOGO·할인형 수신 후 완료율: 55.19 %
BOGO·할인형 열람 후 완료율: 50.69 %
완료 중 사전 열람이 없었던 비율: 28.64 %

[미연결 이벤트 점검]
원본 열람 이벤트: 57725
수신 건에 연결된 열람: 56567
연결되지 않은 열람: 1158
원본 완료 이벤트: 33579
수신 건에 연결된 완료: 33549
연결되지 않은 완료: 30

[중복 의심 기록]
전체 중복 의심 기록: 397


,중복 의심 건수
event,
offer completed,397


## 관찰기간 및 이벤트 연결 검증

### 관찰기간 점검

전체 프로모션 수신 76,277건 중 유효기간 전체가 관찰기간 안에 포함된 수신 건은 66,041건이었다. 나머지 10,236건은 데이터 관찰 종료 시점인 714시간 이후까지 유효기간이 남아 있어 열람 및 완료 기회가 충분하지 않았을 수 있다.

이에 따라 최종 프로모션 성과 비교에는 유효기간 전체가 관찰된 수신 건을 사용한다.

### 관찰기간이 확보된 수신 건 기준 지표

| 지표 | 결과 |
|---|---:|
| 전체 수신 후 열람률 | 74.83% |
| BOGO·할인형 수신 후 열람률 | 77.69% |
| BOGO·할인형 수신 후 완료율 | 55.19% |
| BOGO·할인형 열람 후 완료율 | 50.69% |
| 완료 중 사전 열람이 없었던 비율 | 28.64% |

관찰기간 보정 전후의 차이는 0.16~1.32%p로 비교적 작았다. 관찰기간 부족이 지표를 일부 낮췄지만 전체적인 결과를 크게 바꾸지는 않았다.

수신 후 완료율은 참고값으로만 사용하고, 핵심 지표는 수신 후 열람률과 열람 후 완료율로 설정한다.

### 이벤트 연결 점검

- 원본 열람 이벤트 57,725건 중 56,567건이 수신 건의 최초 열람으로 연결되었다.
- 연결되지 않은 1,158건에는 반복 열람이나 유효기간 밖의 열람이 포함될 수 있다.
- 원본 완료 이벤트 33,579건 중 33,549건이 유효한 수신 건에 연결되었다.
- 연결되지 않은 완료 이벤트는 30건으로 전체 완료 기록에서 차지하는 비중이 매우 작다.

### 중복 의심 기록

동일하게 보이는 기록 397건은 모두 프로모션 완료 이벤트였다. 동일 고객이 동일 프로모션을 반복 수신한 경우 여러 활성 프로모션이 같은 시간에 완료될 가능성이 있으므로, 해당 기록을 즉시 삭제하지 않는다.

반복 수신과 대응되지 않는 완료 기록인지 추가 확인한 후 실제 중복 여부를 결정한다.

### 7. 프로모션 유형별 퍼널 지표 계산

In [9]:
import numpy as np

# 7. 프로모션 유형별 퍼널 지표 계산
offer_type_summary = (
    fully_observed
    .groupby("offer_type")
    .agg(
        수신_건수=("offer_id", "size"),
        열람_건수=("viewed", "sum"),
        완료_건수=("completed", "sum"),
        열람후_완료_건수=("viewed_before_completed", "sum"),
        사전열람없는_완료_건수=(
            "completed_without_prior_view",
            "sum"
        )
    )
)

offer_type_summary["수신후_열람률"] = (
    offer_type_summary["열람_건수"]
    / offer_type_summary["수신_건수"]
    * 100
)

# informational은 완료율 적용 대상이 아니므로 NaN 처리
incentive_type = offer_type_summary.index.isin(
    ["bogo", "discount"]
)

offer_type_summary["열람후_완료율"] = np.where(
    incentive_type,
    (
        offer_type_summary["열람후_완료_건수"]
        / offer_type_summary["열람_건수"]
        * 100
    ),
    np.nan
)

offer_type_summary["사전열람없는_완료_비율"] = np.where(
    incentive_type,
    (
        offer_type_summary["사전열람없는_완료_건수"]
        / offer_type_summary["완료_건수"]
        * 100
    ),
    np.nan
)

rate_columns = [
    "수신후_열람률",
    "열람후_완료율",
    "사전열람없는_완료_비율"
]

offer_type_summary[rate_columns] = (
    offer_type_summary[rate_columns].round(2)
)

display(offer_type_summary)

,수신_건수,열람_건수,완료_건수,열람후_완료_건수,사전열람없는_완료_건수,수신후_열람률,열람후_완료율,사전열람없는_완료_비율
offer_type,,,,,,,,
bogo,27897,23368,14313,10317,3996,83.77,44.15,27.92
discount,22909,16102,13726,9691,4035,70.29,60.19,29.40
informational,15235,9947,0,0,0,65.29,NaN,NaN


## 프로모션 유형별 퍼널 비교

### 분석 결과

| 프로모션 유형 | 수신 건수 | 수신 후 열람률 | 열람 후 완료율 | 사전 열람 없는 완료 비율 |
|---|---:|---:|---:|---:|
| BOGO | 27,897 | 83.77% | 44.15% | 27.92% |
| 할인형 | 22,909 | 70.29% | 60.19% | 29.40% |
| 정보 제공형 | 15,235 | 65.29% | 적용 불가 | 적용 불가 |

### 해석

BOGO는 수신 후 열람률이 83.77%로 세 유형 중 가장 높았지만, 열람 후 완료율은 44.15%로 할인형보다 낮았다.

할인형은 수신 후 열람률이 70.29%로 BOGO보다 낮았지만, 열람 후 완료율은 60.19%로 더 높았다. 따라서 BOGO는 열람 유도, 할인형은 열람 이후 완료 전환에서 상대적으로 높은 성과를 보였다.

정보 제공형은 완료 조건이 없으므로 열람률만 평가했다. 수신 후 열람률은 65.29%로 나타났다.

이 결과는 프로모션 유형과 고객 행동의 연관성을 보여주며, 특정 유형이 성과를 직접 발생시켰다는 인과관계로 해석하지 않는다.

### 8. 개별 프로모션별 퍼널 지표

In [10]:
# 8. 개별 프로모션별 퍼널 지표

offer_detail = fully_observed.copy()

offer_detail["channel_combination"] = (
    offer_detail["channels"]
    .apply(lambda x: ", ".join(x))
)

offer_summary = (
    offer_detail
    .groupby(
        [
            "offer_id",
            "offer_type",
            "reward",
            "difficulty",
            "duration",
            "channel_combination"
        ]
    )
    .agg(
        수신_건수=("offer_id", "size"),
        열람_건수=("viewed", "sum"),
        완료_건수=("completed", "sum"),
        열람후_완료_건수=(
            "viewed_before_completed",
            "sum"
        )
    )
    .reset_index()
)

offer_summary["수신후_열람률"] = (
    offer_summary["열람_건수"]
    / offer_summary["수신_건수"]
    * 100
)

incentive_mask = offer_summary["offer_type"].isin(
    ["bogo", "discount"]
)

offer_summary["열람후_완료율"] = np.where(
    incentive_mask,
    (
        offer_summary["열람후_완료_건수"]
        / offer_summary["열람_건수"]
        * 100
    ),
    np.nan
)

offer_summary[
    ["수신후_열람률", "열람후_완료율"]
] = offer_summary[
    ["수신후_열람률", "열람후_완료율"]
].round(2)

offer_summary = offer_summary.sort_values(
    ["offer_type", "열람후_완료율"],
    ascending=[True, False]
)

display(offer_summary)

,offer_id,offer_type,reward,difficulty,duration,channel_combination,수신_건수,열람_건수,완료_건수,열람후_완료_건수,수신후_열람률,열람후_완료율
6,9b98b8c7a33c4b65b9aebfe6a799e6d9,bogo,5,5,7,"web, email, mobile",6351,3392,3619,1780,53.41,52.48
8,f19421c1d4aa40978ebb69ca19b0e20d,bogo,5,5,5,"web, email, mobile, social",7571,7208,4296,3543,95.21,49.15
7,ae264e3637204a6fb9bb56bc8210ddfd,bogo,10,10,7,"email, mobile, social",6382,5533,3067,2225,86.70,40.21
4,4d5c57ea9a6940dd891ad53e9dbe8da0,bogo,10,10,5,"web, email, mobile, social",7593,7235,3331,2769,95.29,38.27
9,fafdcd668e3743c1bb461111dcafc2a4,discount,2,10,10,"web, email, mobile, social",5033,4869,3625,3222,96.74,66.17
1,2298d6c36e964ae4a3e7e9706d1fb8c2,discount,3,7,7,"web, email, mobile, social",6332,6037,4289,3684,95.34,61.02
0,0b1e1539f2cc45b7b9fa7c272da2e1d7,discount,5,20,10,"web, email",5133,1802,2435,976,35.11,54.16
2,2906b810c7d4411798c6938adc9daaa5,discount,2,10,7,"web, email, mobile",6411,3394,3377,1809,52.94,53.30
3,3f207df678b143eea3cee63160fa8bed,informational,0,0,4,"web, email, mobile",7617,3798,0,0,49.86,NaN
5,5a8bc65990b245e5a138643cd4eb9837,informational,0,0,3,"email, mobile, social",7618,6149,0,0,80.72,NaN


## 개별 프로모션별 퍼널 비교

### 주요 결과

개별 프로모션별로 수신 후 열람률과 열람 후 완료율을 비교한 결과, 열람률이 높은 프로모션이 반드시 완료율도 높은 것은 아니었다.

BOGO 프로모션 중 `4d5c...`는 수신 후 열람률이 95.29%로 가장 높았지만 열람 후 완료율은 38.27%로 가장 낮았다. 반면 `9b98...`는 열람률이 53.41%로 낮았지만 열람 후 완료율은 52.48%로 BOGO 중 가장 높았다.

할인형 `fafd...`는 수신 후 열람률 96.74%, 열람 후 완료율 66.17%로 두 지표 모두 높게 나타났다.

### 해석

프로모션의 열람 성과와 완료 성과는 서로 다를 수 있다. 따라서 전체 완료 건수만 비교하지 않고 수신 후 열람과 열람 후 완료 단계를 분리해서 평가해야 한다.

유형별로는 BOGO가 고객의 열람을 유도하는 데 상대적으로 강하고, 할인형은 열람 이후 완료로 연결하는 데 상대적으로 강한 패턴을 보였다.

BOGO의 낮은 열람 후 완료율은 혜택에 대한 관심과 실제 조건 달성 사이에 차이가 있을 가능성을 보여준다. 다만 고객이 완료하지 않은 이유는 현재 데이터에 없으므로, 추가 구매 부담이나 계획의 어려움은 후속 가설로만 해석한다.

`social`이 포함된 프로모션은 대체로 높은 열람률을 보였지만, 채널·보상·난이도·기간이 함께 다르므로 개별 채널의 독립적인 효과로 단정하지 않는다.

### 9. 프로모션 설계 조건별 퍼널 지표

In [11]:
# 9. 프로모션 설계 조건별 퍼널 지표

design_analysis = fully_observed.copy()

# 가설 트리에서 정한 평균 기준으로 구간 생성
design_analysis["reward_level"] = np.where(
    design_analysis["reward"] > 4.2,
    "평균 초과",
    "평균 이하"
)

design_analysis["duration_level"] = np.where(
    design_analysis["duration"] > 6.5,
    "평균 초과",
    "평균 이하"
)

design_analysis["difficulty_level"] = np.where(
    design_analysis["difficulty"] > 7.7,
    "평균 초과",
    "평균 이하"
)


def summarize_factor(data, column, factor_name):
    # 전체 프로모션의 열람률
    view_summary = (
        data.groupby(column)
        .agg(
            수신_건수=("offer_id", "size"),
            열람_건수=("viewed", "sum")
        )
    )

    view_summary["수신후_열람률"] = (
        view_summary["열람_건수"]
        / view_summary["수신_건수"]
        * 100
    )

    # BOGO·할인형 중 실제 열람된 건의 완료율
    incentive_viewed = data.loc[
        data["offer_type"].isin(["bogo", "discount"])
        & data["viewed"]
    ]

    completion_summary = (
        incentive_viewed.groupby(column)
        .agg(
            열람된_인센티브_건수=("offer_id", "size"),
            열람후_완료_건수=(
                "viewed_before_completed",
                "sum"
            )
        )
    )

    completion_summary["열람후_완료율"] = (
        completion_summary["열람후_완료_건수"]
        / completion_summary["열람된_인센티브_건수"]
        * 100
    )

    result = (
        view_summary
        .join(completion_summary, how="left")
        .reset_index()
        .rename(columns={column: "구분"})
    )

    result.insert(0, "설계요소", factor_name)

    rate_columns = [
        "수신후_열람률",
        "열람후_완료율"
    ]

    result[rate_columns] = (
        result[rate_columns].round(2)
    )

    return result


# 보상·기간·난이도
reward_summary = summarize_factor(
    design_analysis,
    "reward_level",
    "보상"
)

duration_summary = summarize_factor(
    design_analysis,
    "duration_level",
    "유효기간"
)

difficulty_summary = summarize_factor(
    design_analysis,
    "difficulty_level",
    "난이도"
)


# 채널 포함 여부
channel_results = []

for channel in ["web", "email", "mobile", "social"]:
    column_name = f"{channel}_included"

    design_analysis[column_name] = (
        design_analysis["channels"]
        .apply(
            lambda channel_list: (
                "포함"
                if channel in channel_list
                else "미포함"
            )
        )
    )

    channel_result = summarize_factor(
        design_analysis,
        column_name,
        channel
    )

    channel_results.append(channel_result)


# 하나의 표로 결합
design_summary = pd.concat(
    [
        reward_summary,
        duration_summary,
        difficulty_summary,
        *channel_results
    ],
    ignore_index=True
)

display(design_summary)

,설계요소,구분,수신_건수,열람_건수,수신후_열람률,열람된_인센티브_건수,열람후_완료_건수,열람후_완료율
0,보상,평균 이하,33011,24247,73.45,14300,8715,60.94
1,보상,평균 초과,33030,25170,76.20,25170,11293,44.87
2,유효기간,평균 이하,30399,24390,80.23,14443,6312,43.70
3,유효기간,평균 초과,35642,25027,70.22,25027,13696,54.72
4,난이도,평균 이하,35489,26584,74.91,16637,9007,54.14
5,난이도,평균 초과,30552,22833,74.73,22833,11001,48.18
6,web,미포함,14000,11682,83.44,5533,2225,40.21
7,web,포함,52041,37735,72.51,33937,17783,52.40
8,email,포함,66041,49417,74.83,39470,20008,50.69
9,mobile,미포함,5133,1802,35.11,1802,976,54.16


## 프로모션 설계 요소별 비교

| 설계 요소 | 비교 | 수신 후 열람률 | 열람 후 완료율 |
|---|---|---:|---:|
| 보상 | 평균 이하 | 73.45% | 60.94% |
| 보상 | 평균 초과 | 76.20% | 44.87% |
| 유효기간 | 평균 이하 | 80.23% | 43.70% |
| 유효기간 | 평균 초과 | 70.22% | 54.72% |
| 난이도 | 평균 이하 | 74.91% | 54.14% |
| 난이도 | 평균 초과 | 74.73% | 48.18% |
| web | 미포함 | 83.44% | 40.21% |
| web | 포함 | 72.51% | 52.40% |
| mobile | 미포함 | 35.11% | 54.16% |
| mobile | 포함 | 78.18% | 50.53% |
| social | 미포함 | 48.55% | 53.16% |
| social | 포함 | 91.37% | 50.01% |

보상 수준이 높은 프로모션은 열람률이 다소 높았지만 열람 후 완료율은 낮았다. 유효기간이 긴 프로모션은 열람률은 낮았지만 열람 후 완료율은 높았다.

난이도에 따른 열람률 차이는 거의 없었지만, 난이도가 높은 프로모션의 열람 후 완료율은 상대적으로 낮았다.

mobile과 social이 포함된 프로모션은 수신 후 열람률이 높았지만 열람 후 완료율은 더 높지 않았다. 따라서 채널 포함 여부는 주로 열람 단계와 연관된 것으로 보인다.

모든 프로모션에 email이 포함되어 있어 email 포함 여부에 따른 성과 차이는 비교할 수 없다.

각 프로모션은 유형, 보상, 기간, 난이도와 채널 조합이 함께 다르다. 따라서 결과는 설계 요소와 성과의 연관성으로 해석하고, 특정 요소의 독립적인 인과효과로 단정하지 않는다.

### 10. 고객 특성별 퍼널 지표

In [12]:
# 10. 고객 특성별 퍼널 지표

customer_analysis = fully_observed.merge(
    profile.rename(columns={"id": "customer_id"}),
    on="customer_id",
    how="left",
    validate="many_to_one"
)

# 고객 구간 생성
customer_analysis["gender_group"] = (
    customer_analysis["gender"].fillna("정보 없음")
)

customer_analysis["income_group"] = np.where(
    customer_analysis["income"].isna(),
    "정보 없음",
    np.where(
        customer_analysis["income"] >= 80000,
        "8만 달러 이상",
        "8만 달러 미만"
    )
)

customer_analysis["age_group"] = np.where(
    customer_analysis["age"] == 118,
    "정보 없음",
    np.where(
        customer_analysis["age"] >= 55,
        "55세 이상",
        "54세 이하"
    )
)

gender_summary = summarize_factor(
    customer_analysis,
    "gender_group",
    "성별"
)

income_summary = summarize_factor(
    customer_analysis,
    "income_group",
    "소득"
)

age_summary = summarize_factor(
    customer_analysis,
    "age_group",
    "연령"
)

customer_summary = pd.concat(
    [
        gender_summary,
        income_summary,
        age_summary
    ],
    ignore_index=True
)

display(customer_summary)

,설계요소,구분,수신_건수,열람_건수,수신후_열람률,열람된_인센티브_건수,열람후_완료_건수,열람후_완료율
0,성별,F,23746,17704,74.56,14174,8980,63.36
1,성별,M,33006,24296,73.61,19398,9876,50.91
2,성별,O,809,682,84.30,525,344,65.52
3,성별,정보 없음,8480,6735,79.42,5373,808,15.04
4,소득,8만 달러 미만,43188,31956,73.99,25404,13412,52.79
5,소득,8만 달러 이상,14373,10726,74.63,8693,5788,66.58
6,소득,정보 없음,8480,6735,79.42,5373,808,15.04
7,연령,54세 이하,27626,20444,74.00,16208,8702,53.69
8,연령,55세 이상,29935,22238,74.29,17889,10498,58.68
9,연령,정보 없음,8480,6735,79.42,5373,808,15.04


## 고객 특성별 퍼널 비교

### 분석 결과

| 고객 특성 | 구분 | 수신 후 열람률 | 열람 후 완료율 |
|---|---|---:|---:|
| 성별 | 여성 | 74.56% | 63.36% |
| 성별 | 남성 | 73.61% | 50.91% |
| 성별 | 기타 | 84.30% | 65.52% |
| 소득 | 8만 달러 미만 | 73.99% | 52.79% |
| 소득 | 8만 달러 이상 | 74.63% | 66.58% |
| 연령 | 54세 이하 | 74.00% | 53.69% |
| 연령 | 55세 이상 | 74.29% | 58.68% |
| 고객정보 | 정보 없음 | 79.42% | 15.04% |

### 해석

성별·소득·연령 집단 사이에서 수신 후 열람률은 큰 차이가 나타나지 않았다. 반면 열람 후 완료율에서는 차이가 나타났다.

여성의 열람 후 완료율은 63.36%로 남성의 50.91%보다 높았다. 소득 8만 달러 이상 고객의 완료율은 66.58%로 8만 달러 미만 고객의 52.79%보다 높았다. 두 결과는 가설과 같은 방향이다.

55세 이상 고객의 열람 후 완료율은 58.68%로 54세 이하 고객의 53.69%보다 높았다. 이는 연령이 높으면 완료율이 낮을 것이라는 가설과 반대 방향이다.

고객정보가 없는 집단은 열람률이 79.42%로 높았지만 열람 후 완료율은 15.04%로 매우 낮았다. 해당 집단은 성별과 소득이 누락되고 나이가 118세로 기록된 동일 고객들로 구성되어 있어 별도 집단으로 해석한다.

현재 결과는 기술통계에 따른 초기 비교이다. 집단 간 차이가 우연인지 판단하려면 통계 검정이 필요하며, 동일 고객이 여러 프로모션을 수신한 반복 관측이라는 점도 고려해야 한다.

In [13]:
%whos DataFrame

Variable              Type         Data/Info
--------------------------------------------
age_summary           DataFrame    Shape: (3, 8)
channel_result        DataFrame    Shape: (2, 8)
completed_incentive   DataFrame    Shape: (28039, 15)
customer_analysis     DataFrame    Shape: (66041, 22)
customer_summary      DataFrame    Shape: (10, 8)
design_analysis       DataFrame    Shape: (66041, 22)
design_summary        DataFrame    Shape: (13, 8)
difficulty_summary    DataFrame    Shape: (2, 8)
duration_summary      DataFrame    Shape: (2, 8)
fully_observed        DataFrame    Shape: (66041, 15)
gender_summary        DataFrame    Shape: (4, 8)
group                 DataFrame    Shape: (3, 5)
incentive_observed    DataFrame    Shape: (50806, 15)
income_summary        DataFrame    Shape: (3, 8)
offer_detail          DataFrame    Shape: (66041, 16)
offer_events          DataFrame    Shape: (167581, 5)
offer_funnel          DataFrame    Shape: (76277, 15)
offer_id_check        DataFrame    

In [14]:
customer_analysis.to_pickle("customer_analysis.pkl")

print("저장 완료:", customer_analysis.shape)

저장 완료: (66041, 22)
